pkg/apis/autoscaling/register.go

```diff
@@ -52,8 +52,78 @@ const (
 	// TargetAnnotationKey is the annotation to specify what metric value the
 	// PodAutoscaler should attempt to maintain. For example,
 	//   autoscaling.knative.dev/metric: cpu
-	//   autoscaling.knative.dev/target: 75   # target 75% cpu utilization
+	//   autoscaling.knative.dev/target: "75"   # target 75% cpu utilization
 	TargetAnnotationKey = GroupName + "/target"
+	// TargetMin is the minimum allowable target. Values less than
+	// zero don't make sense.
+	TargetMin = 1
+
+	// WindowAnnotationKey is the annotation to specify the time
+	// interval over which to calculate the average metric.  Larger
+	// values result in more smoothing. For example,
+	//   autoscaling.knative.dev/metric: concurrency
+	//   autoscaling.knative.dev/window: "2m"
+	// Only the kpa.autoscaling.knative.dev class autoscaler supports
+	// the window annotation.
+	WindowAnnotationKey = GroupName + "/window"
+	// WindowMinSeconds is the minimum allowable stable autoscaling
+	// window. KPA-class autoscalers calculate the desired replica
+	// count every 2 seconds so the closer the window gets to that
+	// value, the more likely data points will be missed entirely by
+	// the panic window which is smaller than the stable
+	// window. Anything less than 6 second isn't going to work well.
+	WindowMinSeconds = 6
+
+	// PanicWindowPercentageAnnotationKey is the annotation to
+	// specify the time interval over which to calculate the average
+	// metric during a spike. Where a spike is defined as the metric
+	// reaching panic level within the panic window (e.g. panic
+	// mode). Lower values make panic mode more sensitive. Note:
+	// Panic threshold can be overridden with the
+	// PanicThresholdPercentageAnnotationKey. For example,
+	//   autoscaling.knative.dev/panicWindowPercentage: "5.0"
+	//   autoscaling.knative.dev/panicThresholdPercentage: "150.0"
+	// Only the kpa.autoscaling.knative.dev class autoscaler supports
+	// the panicWindowPercentage annotation.
+	// Panic window is specified as a percentag to maintain the
+	// autoscaler's algorithm behavior when only the stable window is
+	// specified. The panic window will change along with the stable
+	// window at the default percentage.
+	PanicWindowPercentageAnnotationKey = GroupName + "/panicWindowPercentage"
+	// PanicWindowPercentageMin is the minimum allowable panic window
+	// percentage. The autoscaler calculates desired replicas every 2
+	// seconds, so a panic window less than 2 seconds will be missing
+	// data points. One percent is a very small ratio and would
+	// require a stable window of at least 3.4 minutes. Anything less
+	// doesn't make sense.
+	PanicWindowPercentageMin = 1.0
+	// PanicWindowPercentageMax is the maximum allowable panic window
+	// percentage. The KPA autoscaler's panic feature allows the
+	// autoscaler to be more responsive over a smaller time scale
+	// when necessary. So the panic window cannot be larger than the
+	// stable window.
+	PanicWindowPercentageMax = 100.0
```

Comment:
I dont understand what the reasoning is for this min/max in configmap for the per-kpa attributes. Is the thinking that an operator can set these values in order to prevent a user from foot-gunning?

Reply:
This is not part of the config map.  It just a global.  It's here so that all the relevant parameters and limits are in once place with good documentation.